Code/RDGeneral/testRDTags.cpp

```diff
@@ -0,0 +1,149 @@
+#include "tags.h"
+#include "Invariant.h"
+
+using namespace RDKit;
+
+template < typename T > std::string to_string( const T& n )
+{
+  std::ostringstream stm ;
+  stm << n ;
+  return stm.str() ;
+}
+
+const int MOD=10000;
+const int COUNT=5000000;
+static std::vector<std::string> nums;
+
+void testTags() {
+  for( int idx=0; idx<MOD; ++idx) {
+    nums.push_back( to_string(idx) );
+  }
+
+
+  RDTags tags;
+  {
+    std::clock_t clock1 = std::clock();
+
+    for( int idx=0; idx<COUNT; ++idx) {
+      int i = idx % MOD;
+      CHECK_INVARIANT(tags.get(nums[i]) == common_properties::MAX+i+1,
+                      "Bad map");
+    }
+    std::clock_t clock2 = std::clock();
+    
+    std::cout << "tag gets: " << (double)(clock2-clock1)/CLOCKS_PER_SEC << " s" << std::endl;
+
+  }
+  {
+    std::clock_t clock1 = std::clock();
+
+    for( int idx=0; idx<COUNT; ++idx) {
+      int i = idx % MOD;
+      CHECK_INVARIANT(tags.get(nums[i]) == common_properties::MAX+i+1,
+                      "Bad map");
+    }
+    std::clock_t clock2 = std::clock();
+    
+    std::cout << "tag gets: " << (double)(clock2-clock1)/CLOCKS_PER_SEC << " s" << std::endl;
+
+  }
+  
+  for( int idx=0; idx<COUNT; ++idx) {
+    int i = idx % MOD;
+    CHECK_INVARIANT(nums[i] == tags.get(common_properties::MAX+i+1),
+                    "Bad lookup");
+  }
+  
+  }
+
+#ifdef RDK_TEST_MULTITHREADED
+namespace {
+void runblock_test(RDTags *tag, std::map<std::string, int> *m) {
+  for( int idx=0; idx<COUNT; ++idx) {
+    int i = idx % MOD;
+    int value = tag->get(nums[i]);
+    (*m)[nums[i]] = value;
+  }
+}
+
+void runblock_time(RDTags *tag) {
+  for( int idx=0; idx<COUNT; ++idx) {
+    int i = idx % MOD;
+    tag->get(nums[i]);
+  }
+}
+}
+
+#include <RDGeneral/BoostStartInclude.h>
+#include <boost/thread.hpp>
+#include <RDGeneral/BoostEndInclude.h>
+void testMultiThread() {
+  RDTags tagmap;
+  boost::thread_group tg;
+  int count = 8;
```

Comment:
```
int count = boost::thread::hardware_concurrency();
```

Let the test fill up the entire multi-core machine to maximize the chance at test failures. :-)